tf/tfprocess.py

```diff
@@ -95,18 +95,25 @@ def init_net(self, next_batch):
         self.x = next_batch[0]  # tf.placeholder(tf.float32, [None, 112, 8*8])
         self.y_ = next_batch[1] # tf.placeholder(tf.float32, [None, 1858])
         self.z_ = next_batch[2] # tf.placeholder(tf.float32, [None, 1])
+        self.q_ = next_batch[3] # tf.placeholder(tf.float32, [None, 1])
         self.batch_norm_count = 0
         self.y_conv, self.z_conv = self.construct_net(self.x)
 
+        # y_ has -1 on illegal moves, flush them to 0 first
+        self.y_ = tf.nn.relu(self.y_)
+
         # Calculate loss on policy head
         cross_entropy = \
             tf.nn.softmax_cross_entropy_with_logits(labels=self.y_,
                                                     logits=self.y_conv)
         self.policy_loss = tf.reduce_mean(cross_entropy)
 
         # Loss on value head
+        q_ratio = self.cfg['training'].get('q_ratio', 0)
+        assert 0 <= q_ratio <= 1
+        target = self.q_ * q_ratio + self.z_ * (1 - q_ratio)
         self.mse_loss = \
-            tf.reduce_mean(tf.squared_difference(self.z_, self.z_conv))
+            tf.reduce_mean(tf.squared_difference(target, self.z_conv))
```

Comment:
I assume these changes don't cause any significant change in training speed. (I would hope not, but after SWA I like to ask...)

Reply:
It's just interpolating two tensors of size (batch_size, 1), should be completely negligible. Didn't seem slower when I ran this for 1k steps).